src/GraphQL/Server.hs

```diff
@@ -318,44 +325,57 @@ type family RunUnionHandlerType m (a :: Type) :: Type where
 class RunUnion m a where
   runUnion :: RunUnionHandlerType m a -> AST.Selection -> m GValue.Object
 
-instance forall m typeName interfaces fields rest.
+instance forall m object rest.
          ( RunUnion m rest
          , MonadIO m
          , MonadThrow m
-         , RunFields m (RunFieldsType m fields)
-         , KnownSymbol typeName
-         ) => RunUnion m (Object typeName interfaces fields :<|> rest) where
-  runUnion (lh :<|> rh) fragment@(AST.SelectionInlineFragment (AST.InlineFragment (AST.NamedType queryTypeName) [] subSelection))
-    | typeName == queryTypeName = do
-        result <- buildResolver @m @(Object typeName interfaces fields) lh subSelection
-        -- TODO: See if we can prevent this from happening at compile time.
-        case GValue.toObject result of
-          Nothing -> panic $ "Expected object as result of union query: " <> show result
-          Just object -> pure object
-    | otherwise = runUnion @m @rest rh fragment
-    where typeName = toS (symbolVal (Proxy :: Proxy typeName))
+         , HasGraph m object
+         , HasObjectDefinition object
+         ) => RunUnion m (object :<|> rest) where
```

Comment:
Replacing "Object .." works in this instance because we only have Objects to stick into unions but in the general case we need to spell out the full `Object name interfaces fields` for the type checker. (not actionable but please don't replace any of the other instances)